debezium-connector-oracle/src/main/java/io/debezium/connector/oracle/logminer/processor/ehcache/EhcacheLogMinerTransactionCache.java

```diff
@@ -0,0 +1,187 @@
+import java.util.TreeSet;
+import java.util.function.Consumer;
+import java.util.function.Function;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+import org.ehcache.Cache;
+
+import io.debezium.connector.oracle.logminer.events.LogMinerEvent;
+import io.debezium.connector.oracle.logminer.processor.AbstractLogMinerTransactionCache;
+
+/**
+ * A concrete implementation of {@link AbstractLogMinerTransactionCache} for Ehcache.
+ *
+ * @author Chris Cranford
+ */
+public class EhcacheLogMinerTransactionCache extends AbstractLogMinerTransactionCache<EhcacheTransaction> {
+
+    private final Cache<String, EhcacheTransaction> transactionCache;
+    private final Cache<String, LogMinerEvent> eventCache;
+
+    // Heap-backed caches for quick access to specific metadata to speed up processing
+    private final Map<String, TreeSet<Integer>> eventIdsByTransactionId = new HashMap<>();
+
+    public EhcacheLogMinerTransactionCache(Cache<String, EhcacheTransaction> transactionCache, Cache<String, LogMinerEvent> eventCache) {
+        this.transactionCache = transactionCache;
+        this.eventCache = eventCache;
+    }
+
+    @Override
+    public EhcacheTransaction getTransaction(String transactionId) {
+        return transactionCache.get(transactionId);
+    }
+
+    @Override
+    public void addTransaction(EhcacheTransaction transaction) {
+        transactionCache.put(transaction.getTransactionId(), transaction);
+        eventIdsByTransactionId.put(transaction.getTransactionId(), new TreeSet<>());
+    }
+
+    @Override
+    public void removeTransaction(EhcacheTransaction transaction) {
+        transactionCache.remove(transaction.getTransactionId());
+    }
+
+    @Override
+    public boolean containsTransaction(String transactionId) {
+        return eventIdsByTransactionId.containsKey(transactionId);
+    }
+
+    @Override
+    public boolean isEmpty() {
+        return eventIdsByTransactionId.isEmpty();
+    }
+
+    @Override
+    public int getTransactionCount() {
+        return eventIdsByTransactionId.size();
+    }
+
+    @Override
+    public <R> R streamTransactionsAndReturn(Function<Stream<EhcacheTransaction>, R> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            return consumer.apply(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void transactions(Consumer<Stream<EhcacheTransaction>> consumer) {
+        try (var stream = StreamSupport.stream(transactionCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getValue));
+        }
+    }
+
+    @Override
+    public void eventKeys(Consumer<Stream<String>> consumer) {
+        try (var stream = StreamSupport.stream(eventCache.spliterator(), false)) {
+            consumer.accept(stream.map(Cache.Entry::getKey));
+        }
+    }
+
+    @Override
+    public Iterator<LogMinerEvent> eventsIterator(EhcacheTransaction transaction) {
```

Comment:
Can we switch this to a consumer based implementation instead of returning the Iterator?  Due to the infinispan impl, we were trying to control the closing of the stream.

Reply:
That should be doable. This is only used in 2 places: Commit and Abandonment.

Reply:
@jeremy-l-ford so for the non-heap implementations, the iterator returned has no underlying resources that needs management fwiw. If I change this, I'll need to likely do this using a `Predicate` because I need to break the loop in the commit handler.  Any issue from your PoV using this style?

Reply:
@Naros - using a predicate should be fine.  There's a removeIf(Predicate<Entry<K, V>> predicate) method that's part of the LogMinerCache that lines up with this.